### ISSUE TITLE
SSRF error occurs on Specific IP

### ISSUE THREAD

**Issue by safa-kajabi** [OP-Engineering/link-preview-js#150]
**Describe the bug**

Recently a user reported that they are unable to display a link preview for their shared link. I investigated this and found that the library throws an error for that specific https://nillaskitchen.com/2024/01/27/steg-for-att-sjalvstandigt-balansera-ibs-eller-andra-problem-mage-tarm/. The issue occurs during the resolveDNSHost part; when the address is 192.0.78.212, it throws the [SSRF error](https://github.com/OP-Engineering/link-preview-js?tab=readme-ov-file#ssrf-concerns). The IP is belongs to [wordpress](http://192.0.78.212/). Do you have any suggestion why it doesn't work or an idea why this occurs for that IP and a way to resolve this issue?

**To Reproduce**

**Expected behavior**

- It should generate link preview for that specific link.

**Screenshots**

<img width="1038" alt="Screenshot 2024-05-09 at 14 16 15" src="https://github.com/OP-Engineering/link-preview-js/assets/127859730/4ac9a460-8588-402c-a5d5-3cc7a34d3220">

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser: Any browser
 - Version: -

**Additional context**

**Comment by ospfranco** [comment]
Interesting, there is some built-in protection against loopback requests, it's tested via REGEX:

```
  REGEX_LOOPBACK: new RegExp(
    "^" +
      "(?:(?:10|127)(?:\\.\\d{1,3}){3})" +
      "|" +
      "(?:(?:169\\.254|192\\.168|192\\.0)(?:\\.\\d{1,3}){2})" +
      "|" +
      "(?:172\\.(?:1[6-9]|2\\d|3[0-1])(?:\\.\\d{1,3}){2})" +
      "$",
    "i"
  ),
```

It seems it is matching the second capture group (192.0) which seems it is not a valid IP address? TBH I'm not a security expert, this was sent by some security company after auditing the package years ago so the regex might be wrong.

**Comment by ospfranco** [comment]
It was introduced here:

https://github.com/OP-Engineering/link-preview-js/pull/117

@aaydin-tr this was a regex you suggested, any reason 192.0 is a loopback address?

**Comment by aaydin-tr** [comment]
Hi, thank you for pointing that out!

At the time, I mistakenly included 192.0.x.x as a loopback address, thinking it was part of the reserved IP ranges. However, you're correct—192.0.x.x isn't a loopback address but rather part of reserved ranges for special purposes like documentation (e.g., 192.0.2.0/24). I'll adjust the regex accordingly to exclude 192.0.x.x from the loopback section.

```ts
const REGEX_LOOPBACK = new RegExp(
  "^" +
    // Loopback: 127.0.0.0 - 127.255.255.255
    "(?:127(?:\\.\\d{1,3}){3})" +
    "|" +
    // Private Class A: 10.0.0.0 - 10.255.255.255
    "(?:10(?:\\.\\d{1,3}){3})" +
    "|" +
    // Private Class B: 172.16.0.0 - 172.31.255.255
    "(?:172\\.(?:1[6-9]|2\\d|3[0-1])(?:\\.\\d{1,3}){2})" +
    "|" +
    // Private Class C: 192.168.0.0 - 192.168.255.255
    "(?:192\\.168(?:\\.\\d{1,3}){2})" +
    "|" +
    // Link-local: 169.254.0.0 - 169.254.255.255
    "(?:169\\.254(?:\\.\\d{1,3}){2})" +
    "|" +
    // Documentation: 192.0.2.0/24, 198.51.100.0/24, 203.0.113.0/24
    "(?:192\\.0\\.2(?:\\.\\d{1,3}){1})" +
    "|" +
    "(?:198\\.51\\.100(?:\\.\\d{1,3}){1})" +
    "|" +
    "(?:203\\.0\\.113(?:\\.\\d{1,3}){1})" +
    "|" +
    // Carrier-Grade NAT (CGNAT): 100.64.0.0 - 100.127.255.255
    "(?:100\\.(?:6[4-9]|[7-9]\\d|1[0-1]\\d)(?:\\.\\d{1,3}){2})" +
    "$",
  "i"
);

```

You can also check official docs https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml

**Comment by ospfranco** [comment]
Thanks a lot @aaydin-tr, will replace and create a new version
